src/main/java/co/cask/http/HttpMethodInfo.java

```diff
@@ -125,19 +127,21 @@ private void bodyConsumerChunk(ChannelBuffer buffer) throws HandlerException {
   private void bodyConsumerFinish() {
     BodyConsumer consumer = bodyConsumer;
     bodyConsumer = null;
-    consumer.finished(responder);
+    try {
+      consumer.finished(responder);
+    } catch (Throwable t) {
+      exceptionHandler.handle(t, request, responder);
+    }
   }
 
   /**
-   * Calls {@link BodyConsumer#handleError(Throwable)} and throws {@link HandlerException}. The current
+   * Calls {@link BodyConsumer#handleError(Throwable)}. The current
    * bodyConsumer will be set to {@code null} after the call.
    */
-  private HandlerException bodyConsumerError(Throwable cause) throws HandlerException {
+  private void bodyConsumerError(Throwable cause) {
     BodyConsumer consumer = bodyConsumer;
     bodyConsumer = null;
     consumer.handleError(cause);
```

Comment:
It would be a good idea to wrap invocation of consumer.handleError() in a try block.